contracts/extensions/CoinMachine.sol

```diff
@@ -0,0 +1,197 @@
+
+  uint256 periodLength; // Duration of a sale period (i.e. one hour)
+  uint256 windowSize; // Number of periods in the price averaging window (i.e. 10)
+
+  uint256 targetPerPeriod; // Target number of tokens to sell in a period
+  uint256 maxPerPeriod; // Maximum number of tokens sellable in a period
+  int256 tokenSurplus; // Total tokens undersold (i.e. below target), negative means deficit
+
+  uint256 activePeriod; // The active sale period
+  uint256 activePrice; // The active sale price
+
+  uint256 tokensSold; // Tokens sold in the active period
+  uint256 activeIntake; // Payment received in the active period
+
+  mapping (uint256 => uint256) pastIntakes; // Payment received in past periods
+
+  // Public
+
+  function initialise(
+    address _purchaseToken,
+    uint256 _periodLength,
+    uint256 _windowSize,
+    uint256 _targetPerPeriod,
+    uint256 _maxPerPeriod,
+    uint256 _startingPrice
+  )
+    public
+  {
+    require(activePeriod == 0, "coin-machine-already-initialised");
+
+    purchaseToken = _purchaseToken;
+
+    periodLength = _periodLength;
+    windowSize = _windowSize;
+
+    targetPerPeriod = _targetPerPeriod;
+    maxPerPeriod = _maxPerPeriod;
+
+    activePrice = _startingPrice;
+    activePeriod = getCurrentPeriod();
+
+    uint256 startingIntake = wmul(targetPerPeriod, _startingPrice);
+    for (uint256 i; i < _windowSize; i++) {
+      pastIntakes[i] = startingIntake;
+    }
+  }
+
+  function buyTokens(uint256 _numTokens) public payable {
+    updatePeriod();
+
+    uint256 numTokens = min(_numTokens, maxPerPeriod - tokensSold);
+    uint256 totalPrice = wmul(numTokens, activePrice);
+
+    activeIntake += totalPrice;
+    tokensSold += numTokens;
+
+    assert(tokensSold <= maxPerPeriod);
+
+    if (purchaseToken == address(0x0)) {
+      require(msg.value >= totalPrice, "coin-machine-insufficient-funds");
+      msg.sender.transfer(msg.value - totalPrice);
+    } else {
+      ERC20Extended(purchaseToken).transferFrom(msg.sender, address(this), totalPrice);
+    }
+
+    colony.mintTokensFor(msg.sender, numTokens);
+  }
+
+  // Make sure this is called at least once during the averaging period
+  function updatePeriod() public {
+    uint256 currentPeriod = getCurrentPeriod();
+
+    // Handle the recently elapsed period
+    if (activePeriod < currentPeriod) {
+      pastIntakes[toBin(activePeriod)] = activeIntake;
+      activeIntake = 0;
+
+      tokenSurplus += (int256(targetPerPeriod) - int256(tokensSold));
+
+      tokensSold = 0;
+      activePeriod += 1;
+    }
+
+    // Handle any additional missed periods
+    while (activePeriod < currentPeriod) {
+      pastIntakes[toBin(activePeriod)] = 0;
+      tokenSurplus += int256(targetPerPeriod);
+      activePeriod += 1;
+    }
+
+    // Update the price
+    uint256 sum;
+
+    for (uint256 i; i < windowSize; i++) {
+      sum += pastIntakes[i];
+    }
+
+    activePrice = wdiv(sum / windowSize, targetPerPeriod);
```

Comment:
Maybe this more concretely expresses my distaste for `wdiv` / `wmul`. Here, we're using `wdiv`, passing in two regular `int`s, and getting a `wad` out - without comment! In fact, I do not believe there is a single use in this contract of either function where two `wad`s are passed in (which is how they are intended to be used).

I'm not suggesting a change here necessarily, just trying to explain further my position :smile:

Reply:
Unfortunately you are mistaken regarding what units are WADs, but that's my fault for not providing sufficient documentation.

Reply:
Although I suppose I could re-organize the math around Wei instead of Ether as units and perhaps then the wad math can be removed.

Reply:
Although if the num tokens is Wei and the price is Wei, then that means the price can never drop below 1:1 (i.e. 1 Wei of price for every 1 Wei of token). So I think keeping it 18-decimal denominated is good, since then the price is more flexible (i.e. 1 WAD of token could cost as little as 1 Wei, and vice versa)

Reply:
I think there probably needs to be some, as regardless of your base unit, you need `0<price<1` being possible, I assume.